File: Client/main/java/com/example/securechat/crypto/KeyExchange.java
package com.example.securechat.crypto;

import org.json.JSONObject;

public abstract class KeyExchange {
    protected String algorithmName;

    public KeyExchange(String algorithmName) {
        this.algorithmName = algorithmName;
    }

    // Abstract methods - must be implemented by subclasses
    public abstract void generatePrivateKey();
    public abstract JSONObject generatePublicKey() throws Exception;
    public abstract void computeSharedSecret(JSONObject serverPublicKey) throws Exception;
    public abstract byte[] getSharedSecretBytes();

    // Optional: Validate public key
    public boolean validatePublicKey(JSONObject publicKey) {
        return true; // Default implementation
    }

    // Get algorithm name
    public String getAlgorithmName() {
        return algorithmName;
    }

    // Get key size in bits
    public abstract int getKeySize();

    // Get algorithm parameters
    public abstract JSONObject getParameters() throws Exception;
}

File: Server/src/crypto/keyExchange.js
export class KeyExchange {
  constructor(algorithmName) {
    if (new.target === KeyExchange) {
      throw new TypeError("Cannot construct KeyExchange instances directly");
    }
    this.algorithmName = algorithmName;
  }

  // Abstract methods - must be implemented by subclasses
  generatePrivateKey() {
    throw new Error("Method 'generatePrivateKey()' must be implemented");
  }

  generatePublicKey(privateKey) {
    throw new Error("Method 'generatePublicKey()' must be implemented");
  }

  computeSharedSecret(privateKey, publicKey) {
    throw new Error("Method 'computeSharedSecret()' must be implemented");
  }

  // Serialize public key to JSON-friendly format
  serializePublicKey(publicKey) {
    throw new Error("Method 'serializePublicKey()' must be implemented");
  }

  // Deserialize public key from JSON format
  deserializePublicKey(data) {
    throw new Error("Method 'deserializePublicKey()' must be implemented");
  }

  // Optional: Validate public key
  validatePublicKey(publicKey) {
    return true; // Default implementation
  }

  // Get algorithm name
  getAlgorithmName() {
    return this.algorithmName;
  }

  // Get key size in bits
  getKeySize() {
    throw new Error("Method 'getKeySize()' must be implemented");
  }

  // Get algorithm parameters as JSON
  getParameters() {
    throw new Error("Method 'getParameters()' must be implemented");
  }
}